src/sage/rings/species.py

```diff
@@ -0,0 +1,2337 @@
+                        [(i, i+1) for i in range(1, n, 2)]]
+                self(PermutationGroup(gens), pi, check=False).rename(f"Pb_{n}" + sort)
+
+    def __contains__(self, x):
+        r"""
+        Return whether ``x`` is in ``self``.
+
+        TESTS::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: A = AtomicSpecies("X")
+            sage: G = PermutationGroup([[(1,2)], [(3,4)]]); G
+            Permutation Group with generators [(3,4), (1,2)]
+            sage: G.disjoint_direct_product_decomposition()
+            {{1, 2}, {3, 4}}
+            sage: G in A
+            False
+
+        For convenience, directly indecomposable permutation groups
+        are regarded as being in `AtomicSpecies`::
+
+            sage: G = PermutationGroup([(1,2)])
+            sage: G in AtomicSpecies("X")
+            True
+            sage: A(G) in A
+            True
+            sage: G in AtomicSpecies("X, Y")
+            False
+            sage: (G, {0: [1,2]}) in AtomicSpecies("X, Y")
+            True
+            sage: (G, {3: [1,2]}) in AtomicSpecies("X, Y")
+            False
+            sage: (G, {0: [1]}) in AtomicSpecies("X, Y")
+            False
+            sage: (G, {0: [1], 1: [2]}) in AtomicSpecies("X, Y")
+            False
+            sage: (0, {0: []}) in AtomicSpecies("X, Y")
+            False
+        """
+        if parent(x) == self:
+            return True
+        if isinstance(x, PermutationGroup_generic):
+            if self._arity == 1:
+                G = x
+                pi = {0: G.domain()}
+            else:
+                return False
+        else:
+            G, pi = x
+        if not isinstance(G, PermutationGroup_generic):
+            return False
+        if not set(pi.keys()).issubset(range(self._arity)):
+            return False
+        if (sum(len(p) for p in pi.values()) != len(G.domain())
+            or set(chain.from_iterable(pi.values())) != set(G.domain())):
+            return False
+        for orbit in G.orbits():
+            if not any(set(orbit).issubset(p) for p in pi.values()):
+                return False
+        return len(G.disjoint_direct_product_decomposition()) <= 1
+
+    def grading_set(self):
+        r"""
+        Return the set of non-negative integer vectors, whose length is
+        the arity of ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: AtomicSpecies(["X"]).grading_set()
+            Integer vectors of length 1
+        """
+        return IntegerVectors(length=self._arity)
+
+    def subset(self, size):
+        """
+        Return the set of atomic species with given total cardinality.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import AtomicSpecies
+            sage: A = AtomicSpecies(["X", "Y"])
+            sage: sorted(A.subset(3))
+            [E_3(X), C_3(X), E_3(Y), C_3(Y)]
+        """
+        result = Set()
```

Comment:
Would it be possible to avoid using Sage's `Set` in favor of Python's `set` or `frozenset`? The former is quite slow compared to the Python implementations.

Reply:
unfortunately, this would make `_test_graded_components` fail.  This test is skipped because we cannot restrict to sensible degrees.  In principle, it wants that the result is in `EnumeratedSets()`.

Reply:
Okay, it comes from using `graded_component`, which is what is required to be an actual category object. So be it.